### PR TITLE
Threshold generalized histogram

### PR DESCRIPTION
**Pull request by divyank0**
## Description

#5189
Feature request:   generalized_histogram_thresholding   algorithms

It contains the required function, doc string and test cases. 
I am working on Gallery examples.   Would love a Review till here. 

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/filters/tests/test_thresholding.py [hunk 1]
```diff
@@ -703,3 +704,29 @@ def test_multiotsu_lut():
...
+def test_theshold_generalized_histogram():
+    """
+    tests for generalized histogram thresholding algorithm
+    """
+
+    image = data.camera()
+    counts, bin_centers = histogram(image.ravel(), 256, source_range="image")
+
+    # (nu, tau, kappa, omega, threshold)
+    possible_values = [
```
Comment (mkcor) [inline skimage/filters/tests/test_thresholding.py:718]: I guess you want to [parametrize](https://docs.pytest.org/en/stable/example/parametrize.html) this test :slightly_smiling_face:

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -1233,3 +1233,141 @@ def threshold_multiotsu(image, classes=3, nbins=256):
...
+    >>> default_nu = np.sum(counts)
+    >>> default_tau = np.sqrt(1/12)
+    >>> default_kappa = np.sum(counts)
+    >>> default_omega = 0.5
+    >>>
+    >>> t, score = theshold_generalized_histogram(counts, bin_centers,
+            default_nu, default_tau, default_kappa, default_omega)
+    >>> binary = data<=t
+    """
+    assert nu >= 0
```
Comment (mkcor) [inline skimage/filters/thresholding.py:1364]: Use assertions to self-test the program, not to handle wrong user input. Instead, raise an exception and print an error message. Possibly handle the exception by overwriting wrong input values with the respective default parameter values?

**Review by mkcor — COMMENTED** [review]
Thank you for your contribution! I just took a quick look at your code. I will review more thoroughly later.

**Comment by sciunto** [comment]
I think we need to discuss why this addition is valuable for the library. A broad range of threshold algorithm exists and we need to focus on the most famous one. The proposed one seems to be new. Is there a clear advantage?

File: skimage/filters/tests/test_thresholding.py (hunk 1 above)
Comment (divyank0) [reply skimage/filters/tests/test_thresholding.py]: done

**Comment by divyank0** [comment]
Sure Francois, there is some initial Description written by Juan on the importance of this Feature. 
https://github.com/scikit-image/scikit-image/issues/5189

	
Also 	
While reading the paper and going through other thresholding based global implementations,
I found GHT to be more powerful,  in comparison to other global algorithms, which only returns 1 particular threshold value, with limited control available to the user. 

**Following points I found useful:**
		
1. GHT with its four hyperparameters allow us choose from a wide variety of thresholds, allowing the user to tune to specific use case. Depicted in the following plots.
2.  GHT doesn't require the histogram to be normalized.
3.  tau, τ  hyperparameter serves a similar purpose as coarsing, blurring the input histogram. 
4. GHT can simulate other algorithms as special case. 
5. Robust to parameter change 
image

Thresholds in image 2 to 6 are calculated using GHT with different hyperparameters, 7th is Otsu's Implementation available in scikit-image. 

My knowledge in this subject is limited as compared to others, Hence I welcome any and all comments.

**Comment by jni** [comment]
@sciunto see #5189 for the motivation. The fact that a number of existing algorithms can be generalised to this method, and that this method when properly tuned can outperform many neural nets, is a clear advantage.

Having said this, @divyank0 the main issue I see is that GHT Otsu and existing Otsu *should match*. If they don't, that suggests that there is a bug in the implementation somewhere. Also, see my comment on the original issue for my thoughts about the API:

> The signature should take an image *or* a histogram, where the histogram can be either just the set of frequencies or the bin centres and the frequencies. See the code for Otsu's thresholding for details. The idea is that you can provide the image (as all thresholding methods do), *or*, if you have a precomputed histogram, you can provide the histogram, which is more efficient.
>
> The parameters should all be accessible and yes, defaulting to Otsu's is probably the most sensible default.

**Comment by divyank0** [comment]
I reviewed the code and found that parameters were not configured correctly. 
Changing the nu value to 1e20 from 1e10, fixed the issue for tau=0.01
Please see images attached for explanation. 

for quick reference: 1e30 =  2**100 ; approx.

image
image
image

**Comment by divyank0** [comment]
any update ? @jni

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
